Content.Server/Language/LanguageSystem.cs

```diff
@@ -0,0 +1,191 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Robust.Shared.Prototypes;
+using Content.Server.Popups;
+using Content.Shared.Actions;
+using Content.Shared.Actions.ActionTypes;
+using Content.Shared.Popups;
+
+namespace Content.Server.Language
+{
+    public sealed class LanguageSystem : EntitySystem
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly SharedActionsSystem _actionsSystem = default!;
+
+        private ISawmill _sawmill = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            _sawmill = Logger.GetSawmill("chat.language");
+
+            SubscribeLocalEvent<LinguisticComponent, ComponentInit>(OnInit);
+            SubscribeLocalEvent<LinguisticComponent, ChangeLanguageActionEvent>(OnChangeLanguage);
+
+            SubscribeLocalEvent<ExtraLanguagesComponent, ComponentStartup>(OnExtraLanguages);
+        }
+
+        public bool TryGetAnyValidSpokenLanguage(EntityUid uid, [NotNullWhen(true)] out LanguagePrototype? language, LinguisticComponent? component = null)
+        {
+            language = null;
+
+            if (!Resolve(uid, ref component))
+                return false;
+
+            foreach (var languageId in component.CanSpeak)
+            {
+                if (!_prototypeManager.TryIndex<LanguagePrototype>(languageId, out var candidate))
+                {
+                    _sawmill.Error($"{ToPrettyString(uid)} has an invalid speakable language: {languageId}");
+                    continue;
+                }
```

Comment:
If possible, guard against inserting a bad language in the first place instead of doing this check. From there you can reduce this to `var languageId = component.CanSpeak.first()`